kolibri/core/tasks/utils.py

```diff
@@ -206,3 +212,142 @@ def db_connection():
         repair_sqlite_db(connection)
 
     return connection
+
+
+class ProgressTracker:
+    def __init__(self, total=100):
+
+        # set default values
+        self.progress = 0
+        self.message = ""
+        self.extra_data = None
+
+        # store provided arguments
+        self.total = total
+
+        # Also check that we are not running Python 2:
+        # https://github.com/learningequality/kolibri/issues/6597
+        if sys.version_info[0] == 2:
+            self.progressbar = None
+        else:
+            # Check that we are executing inside a click context
+            # as we only want to display progress bars from the command line.
+            try:
+                click.get_current_context()
+                # Coerce to an integer for safety, as click uses Python `range` on this
+                # value, which requires an integer argument
+                # N.B. because we are only doing this in Python3, safe to just use int,
+                # as long is Py2 only
+                self.progressbar = click.progressbar(length=int(total), width=0)
+            except RuntimeError:
+                self.progressbar = None
+
+    def update_progress(self, increment, message):
+        self.progress += increment
+
+        if self.progressbar:
+            # Click only enforces integers on the total (because it is implemented assuming a length)
+            self.progressbar.update(increment)
+            if message:
+                self.progressbar.label = message
+
+
+class JobProgressMixin(object):
+    """A mixin with convenience functions for displaying
+    progress to the user, and updating progress on a job.
+
+    If ran from the command line, code here displays a progress bar to the
+    user. If ran asynchronously this mixin sends results through the Job class
+    to the main job storage database. Anyone who knows the task id for the
+    job instance can check the intermediate progress by looking at the Job's progress.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.progresstracker = None
+        self.job = get_current_job()
+        super(JobProgressMixin, self).__init__(*args, **kwargs)
+
+    def update_progress(self, increment=1, message="", extra_data=None):
+        if self.progresstracker:
+            self.progresstracker.update_progress(increment, message)
+        if self.job:
+            self.job.update_progress(
+                self.job.progress + increment, self.job.total_progress
+            )
+            if extra_data and isinstance(extra_data, dict):
+                self.job.update_job_metadata(**extra_data)
```

Comment:
Ah, good catch - I think I meant to add it to the Job class, because it seemed like a useful thing to be able to set directly on the job, but then didn't...

Reply:
Keeping that function at Job class makes sense to me sir